### DISCUSSION TITLE
DAI 2023-2024 - Java IOs benchmarking - Walid Slimani

### DISCUSSION THREAD

**Discussion by SlWa99** [heig-vd-dai-course/heig-vd-dai-course#74] [Show and tell]
Data :
| Size (byte) | Operation                       | Non-Buffered Stream | Buffered Stream   |
|-------------|---------------------------------|---------------------|-------------------|
| 1B          | Write to binary1B.bin           | 0ms / 833700ns      | 0ms / 563000ns    |
| 1B          | Read binary1B.bin               | 0ms / 307600ns      | 0ms / 188900ns    |
| 1B          | Write to file1B.txt             | 1ms / 1019000ns     | 0ms / 542800ns    |
| 1B          | Read file1B.txt                 | 0ms / 270200ns      | 0ms / 147300ns    |
| 1KB         | Write to binary1KB.bin          | 4ms / 4072600ns     | 0ms / 687800ns    |
| 1KB         | Read binary1KB.bin              | 0ms / 321200ns      | 0ms / 278300ns    |
| 1KB         | Write to file1KB.txt            | 2ms / 2783400ns     | 0ms / 955600ns    |
| 1KB         | Read file1KB.txt                | 0ms / 444800ns      | 0ms / 365400ns    |
| 1MB         | Write to binary1MB.bin          | 2925ms / 2925970900ns| 1ms / 1684700ns  |
| 1MB         | Read binary1MB.bin              | 0ms / 489000ns      | 0ms / 346600ns    |
| 1MB         | Write to file1MB.txt            | 96ms / 96546900ns   | 20ms / 20610000ns |
| 1MB         | Read file1MB.txt                | 0ms / 479100ns      | 1ms / 1149800ns   |

Questions :
1) For writing and reading small amounts of data (1 byte), non-buffered streams are the most efficient in terms of elapsed time. For larger data sizes (1 KB and 1 MB), buffered streams are the most efficient for both writing and reading.

2) Non-buffered streams are more efficient for small data sizes because they directly write or read the data to/from the file without any intermediate buffering. For larger data sizes, buffered streams are more efficient because they store data in a buffer before writing it to the file, reducing the number of file system calls and improving overall performance.

3) Binary data is composed of sequences of bytes that can represent any type of data, including non-textual information such as images, audio, or program executables. Text data, on the other hand, is composed of characters from a specific character encoding and is typically human-readable. Binary data is not limited to any specific encoding and can contain any bit patterns, whereas text data follows a specific encoding scheme and is designed for human.

4) A character encoding is a scheme that assigns numeric values to characters and provides a way to represent those characters as binary data. It defines how characters are mapped to bytes in a computer's memory or when stored in a file. Common character encodings include ASCII, UTF-8, and UTF-16. Different encodings can represent the same characters differently, so it's important to use the appropriate encoding to ensure that text data is interpreted correctly.

5) The methodology presented here illustrates the performance differences between buffered and non-buffered streams when reading and writing data of different sizes. Understanding the efficiency of these stream types is crucial when developing applications that involve file I/O operations. Choosing the right type of stream can significantly impact the speed and resource usage of the application, especially when dealing with large files or frequent I/O operations. This knowledge helps developers make informed decisions about stream selection based on their specific use cases and performance requirements.

PS: These answer are generated with Chat-GPT. I will use this content like a resume of this pratical content. If this approach is problematic for you, just tell me and i will write with my own words next time 😄.

Link of my GitHub repository : 
https://github.com/SlWa99/WSI_PraticalContent_JavaIOS

**Comment by ludelafo** [comment]
No problem for me if you use ChatGPT! Just be aware that we cannot allow it during the evaluations and the exam. Great usage of the program to make you summaries. Always have a critical mind with the answers it provides:

> For writing and reading small amounts of data (1 byte), non-buffered streams are the most efficient in terms of elapsed time. For larger data sizes (1 KB and 1 MB), buffered streams are the most efficient for both writing and reading.

While totally true, it is very rare to deal with files that are **always** only made of a few bytes. The statement is correct but our general advice is to always use buffered streams to ensure it works the most efficient in most cases.

For text files, do not forget to specify the charset of the file (while reading and writing!). It is more explicit and you can avoid many issues in the future.

One of the constructors for the `FileReader` et `FileWriter` classes allows to specify the charset to ensure the characters are interpreted correctly by your application: 

- https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/io/FileReader.html
- https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/io/FileWriter.html
